diff --git a/CMakeLists.txt b/CMakeLists.txt
index 889fddf..29d3244 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -475,7 +475,7 @@ endif()
 
 target_include_directories(raw
         PUBLIC
-            $<INSTALL_INTERFACE:libraw>
+            $<INSTALL_INTERFACE:include/libraw>
             $<BUILD_INTERFACE:${LIBRAW_PATH}>)
 
 target_link_libraries(raw PUBLIC ${MATH_LIBRARY})
@@ -608,7 +608,6 @@ if (LIBRAW_INSTALL)
                         ${LIBRAW_PATH}/libraw/libraw_version.h
                         ${CMAKE_CURRENT_BINARY_DIR}/libraw_config.h
             DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/libraw
-            COMPONENT   Devel
            )
 
     # Install Shared binary files.
@@ -617,12 +616,11 @@ if (LIBRAW_INSTALL)
             RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
             LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
             ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+            INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
            )
 
 
-    if(NOT BUILD_SHARED_LIBS AND "${CMAKE_CXX_SIMULATE_ID}" STREQUAL "MSVC")
-        message("ClangCl does not support pdb generation with static libraries")
-    elseif(MSVC)
+    if(BUILD_SHARED_LIBS AND MSVC)
         install(FILES ${PROJECT_BINARY_DIR}/raw.pdb ${PROJECT_BINARY_DIR}/raw_r.pdb
                 DESTINATION ${CMAKE_INSTALL_LIBDIR}
                 CONFIGURATIONS Debug RelWithDebInfo
@@ -664,15 +662,15 @@ if (LIBRAW_INSTALL)
     configure_package_config_file(
         cmake/${PROJECT_NAME}Config.cmake.in
         cmake/${PROJECT_NAME}Config.cmake
-        INSTALL_DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/)
+        INSTALL_DESTINATION "share/${PROJECT_NAME}")
 
     install(FILES
         ${CMAKE_CURRENT_BINARY_DIR}/cmake/${PROJECT_NAME}Config.cmake
         ${CMAKE_CURRENT_BINARY_DIR}/cmake/${PROJECT_NAME}ConfigVersion.cmake
-        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/)
+        DESTINATION "share/${PROJECT_NAME}")
 
     install(EXPORT ${PROJECT_NAME}Targets
-        NAMESPACE libraw:: DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/)
+        NAMESPACE libraw:: DESTINATION "share/${PROJECT_NAME}")
 endif(LIBRAW_INSTALL)
 
 # -- Compile LibRaw Examples --------------------------------------------------------------------------------
